gdal/frmts/mrf/mrf_band.cpp

```diff
@@ -186,6 +251,99 @@ static void *DeflateBlock(buf_mgr &src, size_t extrasize, int flags) {
     return src.buffer;
 }
 
+#if defined(ZSTD_SUPPORT)
+
+// Arange bytes by rank in the input
+static void rankfilter(buf_mgr& src, size_t factor) {
+    if (factor > 1) {
+        std::vector<char> tempb(src.size);
+        char* d = tempb.data();
+        for (size_t j = 0; j < factor; j++)
+            for (size_t i = j; i < src.size; i += factor)
+                *d++ = src.buffer[i];
+        memcpy(src.buffer, tempb.data(), src.size);
+    }
+    // Apply per byte delta
+    auto guard = src.buffer + src.size;
+    char b(0);
+    for (auto p = src.buffer; p < guard; p++) {
+        char temp = *p;
+        *p -= b;
+        b = temp;
+    }
+}
+
+static void derank(buf_mgr& src, size_t factor) {
+    // undo the delta
+    char b(0);
+    auto guard = src.buffer + src.size;
+    for (auto p = src.buffer; p < guard; p++) {
+        b += *p;
+        *p = b;
+    }
+    if (factor > 1) { // undo rank separation
+        std::vector<char> tempb(src.size);
+        char* d = tempb.data();
+        size_t chunk = src.size / factor;
+        for (size_t i = 0; i < chunk; i++)
+            for (size_t j = 0; j < factor; j++)
+                *d++ = src.buffer[chunk * j + i];
+        memcpy(src.buffer, tempb.data(), src.size);
+    }
+}
+
+/*
+* Compress a buffer using zstd, extrasize is the available size in the buffer past the input
+* If ranks > 0, apply the rank filter
+* If the output fits past the data, it uses that area, otherwise it uses a temporary buffer
+* and copies the data over the input on return, returning a pointer to it.
+* The output size is returned in src.size
+* Returns nullptr when compression failed
+*/
+static void* ZstdCompBlock(buf_mgr &src, size_t extrasize, int c_level, ZSTD_CCtx *cctx, size_t ranks)
+{
+    if (ranks && (src.size % ranks) == 0)
+        rankfilter(src, ranks);
+
+    // The buffer pointer we might allocate
+    void* dbuff = nullptr;
+    buf_mgr dst = { src.buffer + src.size, extrasize };
+    // Allocate a temp buffer if there is not sufficient space.
+    // Zstd bound is about (size * 1.004 + 64)
+    if (extrasize < ZSTD_compressBound(src.size)) {
+        dst.size = ZSTD_compressBound(src.size);
+        dbuff = VSIMalloc(dst.size);
+        dst.buffer = (char*)dbuff;
+        if (!dst.buffer)
+            return nullptr;
+    }
+
+    size_t val = cctx ? ZSTD_compressCCtx(cctx, dst.buffer, dst.size, src.buffer, src.size, c_level)
```

Comment:
it seems to me that cctx is never null in the 2 call sites of this function

Reply:
hum unless ZSTD_createCCtx() had failed, which should likely be dealt as an error